### DISCUSSION TITLE
Interactive ARG Alongside Other Figures

### DISCUSSION THREAD

**Discussion by kitchensjn** [kitchensjn/tskit_arg_visualizer#60] [Ideas]
This idea stems from a conversation with Gideon Bradburd. Relating to spatial work, he suggested adding an accompanying map (1D,2D,3D?) alongside the ARG with connected interactivity between the figures. You could hover over an edge in the ARG and it would highlight in the map as well. Plotting an ARG on a map can be confusing to read because there are often a lot of edge crosses. Adding the untangled (as best as possible) ARG alongside could help better understand what is going on within the map. There's a lot that could played with here.

And I don't think that this idea is necessarily limited to maps. If there is a figure that has some sort of relationship between specific edges or nodes in the ARG, then I think it could pair nicely. I wanted to add this as a discussion rather than an issue so that I can add various figure ideas that could be interesting. This alongside with alternative visualizations of the ARG itself (full ARG, time slice, node specific) could lead to a small library of various figure modules.

**Comment by kitchensjn** [comment]
Leaf trace plots and TMRCA plots (see [Hubisz and Siepel 2020](https://link.springer.com/protocol/10.1007/978-1-0716-0199-0_10)) might be interesting to have. I think that these would be relatively simple plots as the ARGs being visualized are often not very complex, but it may help users understand how to read these plots before diving into the complex ARGs that are presented in that paper.

**Comment by kitchensjn** [comment]
Running list of plots that I'm interested in adding in no particular order (this comment will regularly be updated):

- [ ] ARG in geospatial context (using tskit.TreeSequence.tables.individuals.locations)
- [x] Leaf trace across the chromosome
- [ ] TMRCA across the chromosome
- [x] Edge interval spans
- [ ] Mutations/sample sequences
- [ ] Breakpoints along chromosome

**Reply by kitchensjn** [reply]
I will add a check when I've got a rough prototype and have written up a brief paragraph on my thoughts. A check does not mean that this will be added to the package, just that it has discussed in more detail below.

**Comment by kitchensjn** [comment]
<img width="1123" alt="Screenshot 2024-02-11 at 5 52 20 PM" src="https://github.com/kitchensjn/tskit_arg_visualizer/assets/40303683/8dc7b1ee-45d5-4961-8d10-f3322de0f299">

### Leaf traces (left)

These plots were first discussed in the ARGweaver paper and used repeatedly in this [chapter](https://link.springer.com/protocol/10.1007/978-1-0716-0199-0_10). I have attempted to implement them using my own code, so it doesn't match their methods perfectly but should provide a similar interpretation. These plots provide a course view of tree shapes across the chromosome. Each line in the plot corresponds with a sample. For each tree in the ARG, I loop through the samples in their specified tip order and calculate the TMRCA between neighboring samples. The lines (samples) are spaced based on their TMRCA to their neighbor. These distances and the overall ordering of the samples will change across trees, which causes the lines to change positions as we scan from left to right.

**Opinion of plot**: These plots can be difficult to interpret and will change dependent on the sample ordering used. I think that they are best at showing broad patterns over long chromosome sequences with many trees, which isn't really where ARG visualizer shines (it's better at small ARGs). The D3ARG object does not contain measurements like TMRCA, so this plot cannot be created from the current implementation; I instead created the plot from its own custom object. This means that the sample ordering in the leaf trace plot will differ from that in the ARG visualizer (not ideal). Best case would be updating this plot as the sample order in the visualizer changes; this would have to be done in JS or Pyodide. Overall, I don't think that this is the most effective implementation of these plots, and it unfortunately doesn't fit very well with the current visualizer format.

### Edge spans (right)

Edge span plots show the chromosome intervals that each edge corresponds to. It can be used to see how edges change between trees. There may also be ways of using these plots to handle quality control for ARG inference methods.

**Opinion of plot**: Edge span plots can be created directly from the D3ARG object and I think these could be a useful addition alongside the visualizer. Linking the interactivity of the plots seems natural, though I haven't yet attempted this. The size of the blocks will need to be relative to how many blocks there are, otherwise there will be overlap. This one seems promising.

**Comment by kitchensjn** [comment]
Another interesting take on a tree sequence visualization is a carousel that displays three trees at a given time and you can scroll through the trees. For larger trees, you may be only able to display one tree at a time. Since it is only working with trees, D3-force-layout is mostly likely not necessary or helpful. Memory issues may arise for long tree sequences unless there is a way to call build them on the fly (Pyodide?).

**Comment by hyanwong** [comment]
Just to flag up that for very large tree sequences, we have a tool at https://github.com/tskit-dev/tsqc, see @savitakartik 's poster below:

[probgen_2024_v12.pdf](https://github.com/kitchensjn/tskit_arg_visualizer/files/15025792/probgen_2024_v12.pdf)
